### PR TITLE
Add CDF of noncentral beta and F distribution

### PR DESCRIPTION
**Pull request by Sumegh-git**
cc @simonbyrne .

### CODE REVIEW

File: src/betanc.jl [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    if x <= 0.0
+        return 0.0
+    elseif x >= 1.0
+        return 1.0
+    end
+    
+    c = 0.5*lambda
+    #Init series
+
+    beta = logabsgamma(a)[1] + logabsgamma(b)[1] - logabsgamma(a+b)[1]
```
Comment (simonbyrne) [inline src/betanc.jl]: use `logabsbeta` here?

File: src/betanc.jl [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    temp = beta_inc(a,b,x)[1]
+    gx = exp(a*log(x) + b*log(1.0-x) - beta - log(a))
```
Comment (simonbyrne) [inline src/betanc.jl]: use `beta_integrand`?

File: src/betanc.jl [hunk 3]
```diff
@@ -0,0 +1,142 @@
+const errmax = 1e-15
+
+#Compute tail of noncentral Beta distribution
+"""
+	ncbeta_tail(x,a,b,lambda)
+
+Compute tail of the noncentral beta distribution.
```
Comment (simonbyrne) [inline src/betanc.jl]: what values is this valid for, and what series/algorithm is this using?

File: src/betanc.jl [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        temp -= gx
+        gx *= x*(a+b+xj-1.0)/(a+xj)
+        q *= c/xj
+        sumq -= q
+        ax = temp*q
+        ans += ax
+
+        #Check convergence
+        errbd = abs((temp-gx)*sumq)
+        if trunc(Int,xj) > 1000 || errbd < 1e-10
```
Comment (simonbyrne) [inline src/betanc.jl]:
```suggestion
        if xj > 1000 || errbd < 1e-10
```

File: src/betanc.jl [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        return 1.0
+    end
+    c = 0.5*lambda
+    xj = 0.0
+
+    if lambda < 54.0
+        return ncbeta_tail(x,a,b,lambda)
+    else
+        m = trunc(Int, c+0.5)
+        mr = float(m)
```
Comment (simonbyrne) [inline src/betanc.jl]:
```suggestion
        mr = round(c)
        m = trunc(Int, mr)
```

File: src/betanc.jl [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        iterlo = m - trunc(Int, 5.0*sqrt(mr))
+        iterhi = m + trunc(Int, 5.0*sqrt(mr))
+        t = -c + mr*log(c) - logabsgamma(mr + 1.0)[1]
+        q = exp(t)
+        r = q
+        psum = q
+
+        beta = logabsgamma(a+mr)[1] + logabsgamma(b)[1] - logabsgamma(a+mr+b)[1]
+        s1 = (a+mr)*log(x) + b*log(1.0-x) - log(a+mr) - beta
```
Comment (simonbyrne) [inline src/betanc.jl]: again, looks like `beta_integrand`

File: src/betanc.jl [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        m = trunc(Int, c+0.5)
+        mr = float(m)
+        iterlo = m - trunc(Int, 5.0*sqrt(mr))
+        iterhi = m + trunc(Int, 5.0*sqrt(mr))
+        t = -c + mr*log(c) - logabsgamma(mr + 1.0)[1]
+        q = exp(t)
+        r = q
+        psum = q
+
+        beta = logabsgamma(a+mr)[1] + logabsgamma(b)[1] - logabsgamma(a+mr+b)[1]
```
Comment (simonbyrne) [inline src/betanc.jl]: use `logabsbeta`

File: src/betanc.jl [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            q *= iter1/c
+            xj += 1.0
+            gx *= (a + iter1)/(x*(a+b+iter1-1.0))
+            iter1 -= 1
+            temp += gx
+            psum += q
+            sm += q*temp
+        end
+
+        t0 = logabsgamma(a+b)[1] - logabsgamma(a+1.0)[1] - logabsgamma(b)[1]
```
Comment (simonbyrne) [inline src/betanc.jl]: use `logabsbeta`

File: src/betanc.jl [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        s = 0.0
+        for i = 1:iter1
+        j = i - 1
+        s += exp(t0+s0+j*log(x))
+        t1 = log(a+b+j) - log(a+j+1.0) + t0
+        t0 = t1
+        end
+        #Compute first part of error bound
+
+        errbd = (1.0 - gamma_inc(float(iter1),c,0)[1])*(temp+s)
```
Comment (simonbyrne) [inline src/betanc.jl]:
```suggestion
        errbd = (gamma_inc(float(iter1),c,0)[2])*(temp+s)
```

File: src/betanc.jl [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        return 0.0
+    elseif x >= 1.0
+        return 1.0
+    end
+    c = 0.5*lambda
+    xj = 0.0
+
+    if lambda < 54.0
+        return ncbeta_tail(x,a,b,lambda)
+    else
```
Comment (simonbyrne) [inline src/betanc.jl]: split this branch into a separate sub-function, and document algorithm/series.

File: src/betanc.jl [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+            temp += gx
+            psum += q
+            sm += q*temp
+        end
+
+        t0 = logabsgamma(a+b)[1] - logabsgamma(a+1.0)[1] - logabsgamma(b)[1]
+        s0 = a*log(x) + b*log(1.0-x)
+
+        s = 0.0
+        for i = 1:iter1
```
Comment (simonbyrne) [inline src/betanc.jl]:
```suggestion
        for j = 0:iter1-1
```

File: src/betanc.jl [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        j = i - 1
```
Comment (simonbyrne) [inline src/betanc.jl]:
```suggestion
```

**Comment by Sumegh-git** [comment]
Done @simonbyrne

File: src/betanc.jl [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    elseif x >= 1.0
+        return 1.0
+    end
+    
+    c = 0.5*lambda
+    #Init series
+
+    beta = logabsbeta(a,b)[1]
+    temp = beta_inc(a,b,x)[1]
+    gx = (SpecialFunctions.beta_integrand(a,b,x,1.0-x))/a
```
Comment (simonbyrne) [inline src/betanc.jl]: `SpecialFunctions` shouldn't be needed here

File: src/betanc.jl [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+	ncbeta_tail(x,a,b,lambda)
+
+Compute tail of the noncentral beta distribution.
+Uses the recursive relation 
+```math
+I_{x}(a,b+1;0) = I_{x}(a,b;0) - \\Gamma(a+b)/\\Gamma(a+1)\\Gamma(b)x^{a}(1-x)^{b}
+```
+and ``\\Gamma(a+1) = a\\Gamma(a)`` given in Equation (26.5.16) in Abromowitz and Stegun(1965).
+"""
+function ncbeta_tail(x::Float64, a::Float64, b::Float64, lambda::Float64)
```
Comment (simonbyrne) [inline src/betanc.jl]: should match the same argument order as `beta_inc`

File: src/betanc.jl [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        ax = temp*q
+        ans += ax
+
+        #Check convergence
+        errbd = abs((temp-gx)*sumq)
+        if xj > 1000 || errbd < 1e-10
+            break
+        end
+    end
+    return ans
```
Comment (simonbyrne) [inline src/betanc.jl]: can we compute both lower and upper tails?

File: src/betanc.jl [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    elseif x >= 1.0
+        return 1.0
+    end
+    c = 0.5*lambda
+    xj = 0.0
+
+    if lambda < 54.0
+        return ncbeta_tail(x,a,b,lambda)
+    else
+        m = trunc(Int, c+0.5)
```
Comment (simonbyrne) [inline src/betanc.jl]:
```suggestion
        m = round(Int, c)
```

File: src/betanc.jl [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+        temp = beta_inc(a+mr,b,x)[1]
+        ftemp = temp
+        xj += 1.0
+
+        sm = q*temp
+        iter1 = m
+
+        #Iterations start from M and goes downwards
+
+        while true
```
Comment (simonbyrne) [inline src/betanc.jl]:
```suggestion
        while !(iter1 < iterlo || q < errmax)
```

File: src/betanc.jl (hunk 15 above)
Comment (Sumegh-git) [reply src/betanc.jl]: Is that needed? Like the paper hasn't written anything on computing upper and lower tails.

Comment (Sumegh-git) [reply src/betanc.jl]: Rather this approach uses a 2-way iteration from middle to top / middle to bottom to improve upon the error bound as mentioned in Documentation.

File: src/betanc.jl (hunk 10 above)
Comment (Sumegh-git) [reply src/betanc.jl]: You mean keeping the `else` part in separate function right? What should I name  it?

File: src/betanc.jl [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+        mr = float(m)
+        iterlo = m - trunc(Int, 5.0*sqrt(mr))
+        iterhi = m + trunc(Int, 5.0*sqrt(mr))
+        t = -c + mr*log(c) - logabsgamma(mr + 1.0)[1]
+        q = exp(t)
+        r = q
+        psum = q
+
+        beta = logabsbeta(a+mr,b)[1]
+        gx = SpecialFunctions.beta_integrand(a+mr,b,x,1.0-x)/(a + mr)
```
Comment (simonbyrne) [inline src/betanc.jl]: `SpecialFunctions` shouldn't be necessary here.

Comment (Sumegh-git) [reply src/betanc.jl]: Sorry! Forgot to remove this one.

File: src/betanc.jl (hunk 15 above)
Comment (simonbyrne) [reply src/betanc.jl]: Like the incomplete beta, this is used for computing the CDF of a probability distribution. If we're computing one tail, we might as well compute the other.

Comment (Sumegh-git) [reply src/betanc.jl]: Yes but in what way? 
This paper has only mentioned for `lambda < 54`.

File: src/betanc.jl [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+#Russell Lenth, Algorithm AS 226: Computing Noncentral Beta Probabilities,
+#Applied Statistics,Volume 36, Number 2, 1987, pages 241-244
+
+"""
+	ncbeta_tail(x,a,b,lambda)
+
+Compute tail of the noncentral beta distribution.
+Uses the recursive relation 
+```math
+I_{x}(a,b+1;0) = I_{x}(a,b;0) - \\Gamma(a+b)/\\Gamma(a+1)\\Gamma(b)x^{a}(1-x)^{b}
```
Comment (simonbyrne) [inline src/betanc.jl]: link to https://dlmf.nist.gov/8.17.21

File: src/betanc.jl [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+```
+and ``\\Gamma(a+1) = a\\Gamma(a)`` given in Equation (26.5.16) in Abromowitz and Stegun(1965).
```
Comment (simonbyrne) [inline src/betanc.jl]: Change link to https://dlmf.nist.gov/8.17.21

File: src/betanc.jl [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+
+#R Chattamvelli, R Shanmugam, Algorithm AS 310: Computing the Non-central Beta Distribution Function,
+#Applied Statistics, Volume 46, Number 1, 1997, pages 146-156
+
+"""
+	ncbeta(a,b,lambda,x)
+
+Compute the CDF of the noncentral beta distribution given by
+```math
+I_{x}(a,b+1;\\lambda ) = \\sum_{j=0}^{\\infty}q(\\lambda/2,j)I_{x}(a+j,b;0)
```
Comment (simonbyrne) [inline src/betanc.jl]: should this be
```suggestion
I_{x}(a,b;\\lambda ) = \\sum_{j=0}^{\\infty}q(\\lambda/2,j)I_{x}(a+j,b;0)
```
?

File: src/betanc.jl (hunk 15 above)
Comment (simonbyrne) [reply src/betanc.jl]: okay, lets leave it for the time being.

**Comment by Sumegh-git** [comment]
@simonbyrne can have a look now.

**Comment by simonbyrne** [comment]
Okay, I think this looks good. Thanks!

**Pull request merged**